tasks/broad/InternalImputationTasks.wdl

```diff
@@ -0,0 +1,128 @@
+version 1.0
+
+task FormatImputationOutputs {
+    input {
+        String          aggregated_imputation_metrics
+        String          chunks_info
+        String          failed_chunks
+        String          imputed_multisample_vcf
+        String          imputed_multisample_vcf_index
+        Array[String]?  imputed_single_sample_vcfs
+        Array[String]?  imputed_single_sample_vcf_indices
+        Array[String]   chip_well_barcodes
+        String          n_failed_chunks
+    }
+
+    command <<<
+
+        # write header to file
+        echo -e "aggregated_imputation_metrics\tchunks_info\tfailed_chunks\tn_failed_chunks\t\
+        imputed_multisample_vcf\timputed_multisample_vcf_index\t\
+        imputed_single_sample_vcfs\timputed_single_sample_vcf_indices\t\
+        chip_well_barcodes" \
+        > ingestDataset_imputation_outputs.tsv
+
+        # handle array[type] variables to print as list with double quotes
+        imputed_single_sample_vcfs='~{sep='","' imputed_single_sample_vcfs}'
+        imputed_single_sample_vcf_indices='~{sep='","' imputed_single_sample_vcf_indices}'
+        chip_well_barcodes='~{sep='","' chip_well_barcodes}'
+
+        # write file paths to row in tsv file
+        echo -e "~{aggregated_imputation_metrics}\t~{chunks_info}\t~{failed_chunks}\t~{n_failed_chunks}\t\
+        ~{imputed_multisample_vcf}\t~{imputed_multisample_vcf_index}\t\
+        [\"${imputed_single_sample_vcfs}\"]\t\
+        [\"${imputed_single_sample_vcf_indices}\"]\t\
+        [\"${chip_well_barcodes}\"]" \
+        >> ingestDataset_imputation_outputs.tsv
+
+
+        python3 << CODE
+        import pandas as pd
+
+        tsv_df = pd.read_csv("ingestDataset_imputation_outputs.tsv", sep="\t")
+        tsv_df = tsv_df.dropna(axis=1, how="all")  # drop columns if no value (optional outputs etc)
+
+        outputs = tsv_df.to_json("ingestDataset_imputation_outputs.json", orient="records")  # write json file
+
+        CODE
+    >>>
+
+    runtime {
+        docker: "broadinstitute/horsefish"
+    }
+
+    output {
+        File ingest_outputs_tsv = "ingestDataset_imputation_outputs.tsv"
+        File ingest_outputs_json = "ingestDataset_imputation_outputs.json"
+    }
+}
+
+task FormatImputationWideOutputs{
+    input {
+        Array[String]?  imputed_single_sample_vcfs
+        Array[String]?  imputed_single_sample_vcf_indices
+    }
+
+    String prefix="[\""
+    String postfix="\"]"
+
+    command <<<
+
+        python3 << CODE
+        import pandas as pd
+        import os
+
+        print("imputed_single_sample_vcfs")
+        single_sample_vcfs=~{prefix}~{sep="\", \"" imputed_single_sample_vcfs}~{postfix}
```

Comment:
I feel like I've done this type of thing in the past as something like `single_sample_vcfs=["~{sep='", ' imputed_single_sample_vcfs}"]`,  the important point being to use single quotes to specify the `sep`, so you don't need to escape the double quotes (or the other way around, I'd think either should work).  But if that doesn't work I don't have a better solution.

Reply:
wait what are we actually trying to do here?

Reply:
the intended output should be 
`["gs://path_to_file/file_1.vcf", "gs://path_to_file/file_2.vcf", "gs://path_to_file/file_3.vcf"]` where the input is an `Array[String]` (coerced to such from `Array[File]`).

I swear I had tried what Chris suggested above as well as a ton of other options and I was always missing one set of quotes somewhere or had quotes in all the right places but one set would be single/double etc. The setting prefix and postfix was the cleanest way - atleast to my brain and eye - to make it work in the end.

The reason it has to be in the format above is because thats the input format to the script in the next task that ingests into TDR.

Reply:
Ah its easier to do this with python instead of WDL, I think i've done something similar here -> https://github.com/broadinstitute/warp/blob/243176810a5923b7599f79eb2aae6536f2d76692/tasks/broad/Utilities.wdl#L259

Reply:
genius. will test.

Reply:
Nice use of a list comprehension to filter out the empty string if there's nothing in the array, that's very cool.

Reply:
that was one thing that i though was weird with wdl to python conversion, if the list was not provided it would coerce it to `[""]`

Reply:
Yeah, I feel like what you really need to be able to handle that edge case nicely in pure WDL is the ability to specify prefix, sep, and suffix (does suffix even exist in WDL?) all together.  Otherwise I think you need a terrible sequence of ifelse's if you don't have access to a more powerful language like python.